File: sample/src/main/swift/Sources/SampleAppCore/Int16Test.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public enum Int16Enum: Int16, Codable {
    case one
    case two
    case three
}

public struct Int16OptionsSet: OptionSet, Codable {

    public let rawValue: Int16

    public init(rawValue: Int16) {
        self.rawValue = rawValue
    }

    static let one = Int16OptionsSet(rawValue: 1 << 0)
    static let two = Int16OptionsSet(rawValue: 1 << 1)
    static let three = Int16OptionsSet(rawValue: 1 << 2)
}

public struct Int16TestStruct: Codable, Hashable {
    public var zero: Int16 = Int16.zero
    public var max: Int16 = Int16.max
    public var min: Int16 = Int16.min
    public var optional: Int16? = Int16.zero
    public var optionalNil: Int16? = nil
}

public protocol Int16TestParamProtocol {
    func testParam(_ param: Int16) -> Bool
}

public protocol Int16TestReturnTypeProtocol {
    func testReturnType() -> Int16
}

public protocol Int16TestOptionalParamProtocol {
    func testOptionalParam(_ param: Int16?) -> Bool
}

public protocol Int16TestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> Int16?
}

public typealias Int16Block = (_ value: Int16) -> Int16
public typealias OptionalInt16Block = (_ value: Int16?) -> Int16?

public class Int16Test {

    public static func testZero() -> Int16 {
        return 0
    }

    public static func testMin() -> Int16 {
        return Int16.min
    }

    public static func testMax() -> Int16 {
        return Int16.max
    }

    public static func testParam(_ param: Int16) -> Bool {
        return param == Int16.max
    }

    public static func testReturnType() -> Int16 {
        return Int16.max
    }

    public static func testOptionalParam(_ param: Int16?) -> Bool {
        return param == Int16.max
    }

    public static func testOptionalReturnType() -> Int16? {
        return Int16.max
    }

    public static func testProtocolParam(_ callback: Int16TestParamProtocol) -> Bool {
        return callback.testParam(Int16.max)
    }

    public static func testProtocolReturnType(_ callback: Int16TestReturnTypeProtocol) -> Int16 {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: Int16TestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(Int16.max)
    }

    public static func testProtocolOptionalReturnType(_ callback: Int16TestOptionalReturnTypeProtocol) -> Int16? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  Int16TestStruct {
        return Int16TestStruct()
    }

    public static func testDecode(_ value: Int16TestStruct) -> Bool {
        return value == Int16TestStruct()
    }

    public static func testEnumEncode(_ rawValue: Int16) -> Int16Enum {
        switch rawValue {
        case Int16Enum.one.rawValue: return Int16Enum.one
        case Int16Enum.two.rawValue: return Int16Enum.two
        case Int16Enum.three.rawValue: return Int16Enum.three
        default: fatalError("Can't find enum with rawValue \(rawValue)")
        }
    }

    public static func testEnumDecode(_ enumValue: Int16Enum) -> Int16 {
        return enumValue.rawValue
    }

    public static func testOptionSetEncode(_ rawValue: Int16) -> Int16OptionsSet {
        return Int16OptionsSet(rawValue: rawValue)
    }

    public static func testOptionSetDecode(_ optionSet: Int16OptionsSet) -> Int16 {
        return optionSet.rawValue
    }

    public static func testBlock(_ block: Int16Block) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalInt16Block) -> Bool {
        let value = block(nil)
        return value == nil
    }

}

File: sample/src/main/java/com/readdle/swiftjava/sample/Int16Test.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

@SwiftValue
enum class Int16Enum(val rawValue: Short) {

    ONE(0), TWO(1), THREE(2);

    companion object {

        private val values = HashMap<Short, Int16Enum>()

        @JvmStatic
        fun valueOf(rawValue: Short): Int16Enum {
            return values[rawValue]!!
        }

        init {
            enumValues<Int16Enum>().forEach {
                values[it.rawValue] = it
            }
        }
    }

}

@SwiftValue
data class Int16OptionsSet(val rawValue: Short = 0) {
    companion object {
        @JvmStatic
        val  one = Int16OptionsSet(1)
        @JvmStatic
        val  two = Int16OptionsSet(2)
        @JvmStatic
        val  three = Int16OptionsSet(4)
    }
}

@SwiftValue
data class Int16TestStruct(var zero: Short = 0,
                           var max: Short = Short.MAX_VALUE,
                           var min: Short = Short.MIN_VALUE,
                           var optional: Short? = 0,
                           var optionalNil: Short? = null)

@SwiftReference
class Int16Test private constructor() {

    @SwiftDelegate(protocols = ["Int16TestParamProtocol"])
    interface Int16ParamProtocol {
        @SwiftCallbackFunc
        fun testParam(param: Short): Boolean
    }

    @SwiftDelegate(protocols = ["Int16TestReturnTypeProtocol"])
    interface Int16ReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testReturnType(): Short
    }

    @SwiftDelegate(protocols = ["Int16TestOptionalParamProtocol"])
    interface Int16OptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(param: Short?): Boolean
    }

    @SwiftDelegate(protocols = ["Int16TestOptionalReturnTypeProtocol"])
    interface Int16OptionalReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testOptionalReturnType(): Short?
    }

    @SwiftBlock("(Int16) -> Int16")
    interface Int16Block {
        fun call(value: Short): Short
    }

    @SwiftBlock("(Int16?) -> Int16?")
    interface OptionalInt16Block {
        fun call(value: Short?): Short?
    }

    companion object {
        @JvmStatic
        external fun testZero(): Short

        @JvmStatic
        external fun testMin(): Short

        @JvmStatic
        external fun testMax(): Short

        @JvmStatic
        external fun testParam(param: Short): Boolean

        @JvmStatic
        external fun testReturnType(): Short

        @JvmStatic
        external fun testOptionalParam(param: Short?): Boolean

        @JvmStatic
        external fun testOptionalReturnType(): Short?

        @JvmStatic
        external fun testProtocolParam(callback: Int16ParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: Int16ReturnTypeProtocol): Short

        @JvmStatic
        external fun testProtocolOptionalParam(callback: Int16OptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: Int16OptionalReturnTypeProtocol): Short?

        @JvmStatic
        external fun testEncode(): Int16TestStruct

        @JvmStatic
        external fun testDecode(value: Int16TestStruct): Boolean
        
        @JvmStatic
        external fun testEnumEncode(rawValue: Short) : Int16Enum

        @JvmStatic
        external fun testEnumDecode(enum: Int16Enum) : Short

        @JvmStatic
        external fun testOptionSetEncode(rawValue: Short) : Int16OptionsSet

        @JvmStatic
        external fun testOptionSetDecode(enum: Int16OptionsSet) : Short
        
        @JvmStatic
        external fun testBlock(@SwiftBlock block: Int16Block): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalInt16Block): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}